keps/sig-auth/266-kubelet-client-certificate-bootstrap-rotation/README.md

```diff
@@ -0,0 +1,156 @@
+# KEP-266: Kubelet client certificate bootstrap and rotation
+
+<!-- toc -->
+- [Release Signoff Checklist](#release-signoff-checklist)
+- [Summary](#summary)
+  - [Goals](#goals)
+- [Proposal](#proposal)
+  - [During Kubelet Boot Sequence](#during-kubelet-boot-sequence)
+  - [As Expiration Approaches](#as-expiration-approaches)
+  - [Certificate Approval](#certificate-approval)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Implementation History](#implementation-history)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+- [x] Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [x] KEP approvers have approved the KEP status as `implementable`
+- [x] Design details are appropriately documented
+- [x] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [x] Graduation criteria is in place
+- [x] "Implementation History" section is up-to-date for milestone
+- [x] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+- [x] User-facing documentation has been created at https://kubernetes.io/docs/tasks/tls/certificate-rotation/
+
+## Summary
+
+Currently, a kubelet has a certificate/key pair that authenticates the kubelet to the kube-apiserver.
+The certificate is supplied to the kubelet when it is first booted, via an out of cluster mechanism.
+This proposal covers a process for obtaining the initial cert/key pair and rotating it as expiration 
+of the certificate approaches.
+
+### Goals
+
+1. Enable exchange of a short-lived credential for a client certificate on a new node when starting a Kubelet for the first time.
+2. Enable renewal/rotation of a Kubelet certificate as it approaches expiration.
+
+## Proposal
+
+### During Kubelet Boot Sequence
+
+1. Look on disk for an existing cert/key pair managed by the certificate manager.
+1. If there is an existing cert/key pair, load them.
+1. If there is no existing cert/key pair, look for cert/key data specified in the kubelet config file:
+    - values encoded in the kubeconfig (CertData, KeyData, CAData)
+    - file references in the kubeconfig (CertFile, KeyFile, CAFile)
+1. If the certificate is a bootstrap certificate, use the certificate to:
+    - generate a key
+    - create a certificate signing request
+    - request a signed certificate from the API server
+    - wait for a response
+    - Store the new cert/key pairs in the Kubelet's certificate directory
```

Comment:
There's a ton of nuance in here as we settled on this, I don't really want to dredge it all up, but we can do a follow up to add some details around how the on disk structure is set up, how failure modes are tested (this is where the bulk of the effort ended up being), and what the kubelet does after a certain period of being unable to get a new key (it restarts after 5m).

Reply:
sounds good, I'll create an issue to track

Reply:
done